web-ui/src/types/ddl.tsx

```diff
@@ -6,24 +6,293 @@
+            return first
+          } else if (parts.length === 2 && scale > 0) {
+            return `${first}.${parts[1].substring(0, Math.min(precision - first.length, scale))}`
+          }
+        } else {
+          return '0'
+        }
+      }
+    })
+    .with({ type: 'TIME' }, () => {
+      return (value: any) => {
+        // We represent TIME as string for now because the data-grid doesn't
+        // have native support for times yet.
+        // See also
+        return dayjs(value).format('HH:mm:ss')
+      }
+    })
+    .with({ type: 'DATE' }, () => {
+      return (value: any) => {
+        return dayjs(value)
+      }
+    })
+    .with({ type: 'TIMESTAMP' }, () => {
+      return (value: any) => {
+        return dayjs(value)
+      }
+    })
+    .with({ type: 'ARRAY' }, () => {
+      return (value: any) => {
+        return value.map((v: any) => {
+          assert(columntype.component !== null && columntype.component !== undefined)
+          return getValueParser(columntype.component)(v)
+        })
+      }
+    })
+    .otherwise(() => {
+      return (value: any) => value
+    })
+}
+
+// Convert a row of strings to an object of typed values.
+//
+// The type conversion is important because for sending a row as an anchor later
+// it needs to have proper types (a number can't be a string etc.)
+export function csvLineToRow(relation: Relation, row: any[]): Row {
+  const genId = Number(row[0])
+  const weight = Number(row[row.length - 1])
+  const records = row.slice(1, row.length - 1)
+
+  const record: Record<string, any> = {}
+  relation.fields.forEach((field, i) => {
+    record[field.name] = parseValueSafe(field.columntype, records[i])
+  })
+
+  return {
+    genId,
+    weight,
+    record
+  }
+}
+
+// We convert fields to a tuple so that we can use it as an anchor in the REST
+// API.
+export function rowToAnchor(relation: Relation, obj: Row): any[] {
+  const tuple: any[] = []
+  console.log(obj, relation)
+  relation.fields.map((col: Field, i: number) => {
+    tuple[i] = obj.record[col.name]
+  })
+
+  return tuple
+}
+
+// Walk the type tree and find the base type.
+//
+// e.g., for a `VARCHAR ARRAY` type, this will return VARCHAR.
+export const findBaseType = (type: ColumnType): ColumnType => {
+  if (type.component) {
+    return findBaseType(type.component)
+  }
+
+  return type
+}
+
+// Parse a value for a SQL type. This is used for parsing values from the
+// backend directly that can be trusted to satisfy the type constraints.
+//
 // TODO: Note that all Number in JS are 64-bit floating point numbers. When we
 // switch to the JSON format we will just parse everything as BigInt which has
 // arbitrary precision so we don't loose anything when displaying numbers from
 // dbsp.
-export const parseSqlType = (sqlType: Field, value: string) =>
-  match(sqlType.columntype)
-    .with({ type: 'BOOLEAN' }, () => Boolean(value))
+export const parseValueSafe = (sqlType: ColumnType, value: string) =>
```

Comment:
you really like functions. I like classes.
I would have created a SqlValue class and subclasses to handle this kind of stuff
But I guess this works.

Reply:
I like classes too -- but I'm not sure/very confident on how well classes work in with the (functional) react framework

Reply:
I am pretty sure that they should work fine. 
I am suggesting this only for stuff that React doesn't call directly.